File: turms-client-swift/Sources/TurmsClient/Service/NotificationService.swift
import Foundation

public class NotificationService {
    private unowned var turmsClient: TurmsClient
    public var notificationListeners: [(Notification) -> Void] = []

    init(_ turmsClient: TurmsClient) {
        self.turmsClient = turmsClient
        self.turmsClient.driver
            .addNotificationListener { [weak self] notification in
                guard notification.hasRelayedRequest, !notification.hasCloseStatus else {
                    return
                }
                if case .createMessageRequest = notification.relayedRequest.kind {
                    return
                }
                let n = Notification(timestamp: Date(timeIntervalSince1970: TimeInterval(notification.timestamp)), requesterId: notification.requesterID, relayedRequest: notification.relayedRequest)
                self?.notificationListeners.forEach {
                    $0(n)
                }
            }
    }

    /// Add a notification listener to receive notifications.
    /// Note: This listener will receive all kinds of notifications excluding messages.
    /// To listen to messages, use ``MessageService/addMessageListener`` instead.
    public func addNotificationListener(listener: @escaping (Notification) -> Void) {
        notificationListeners.append(listener)
    }
}

File: turms-client-dart/lib/src/service/notification_service.dart
import '../../turms_client.dart';

typedef NotificationListener = void Function(Notification notification);

class NotificationService {
  final TurmsClient _turmsClient;
  final List<NotificationListener> _notificationListeners = [];

  NotificationService(this._turmsClient) {
    _turmsClient.driver.addNotificationListener((notification) {
      final isBusinessNotification = notification.hasRelayedRequest() &&
          !notification.relayedRequest.hasCreateMessageRequest() &&
          !notification.hasCloseStatus();
      if (isBusinessNotification) {
        for (final listener in _notificationListeners) {
          listener.call(Notification(
              DateTime.fromMillisecondsSinceEpoch(
                  notification.timestamp.toInt()),
              notification.requesterId,
              notification.relayedRequest));
        }
      }
    });
  }

  /// Add a notification listener to receive notifications.
  /// Note: This listener will receive all kinds of notifications excluding messages.
  /// To listen to messages, use [MessageService.addMessageListener] instead.
  void addNotificationListener(NotificationListener listener) {
    _notificationListeners.add(listener);
  }

  /// Remove a notification listener.
  void removeNotificationListener(NotificationListener listener) {
    _notificationListeners.remove(listener);
  }
}
